### PR TITLE
[BABEL-3286] Support for FORMAT()

### PR DESCRIPTION
**Pull request by timchang514**
Signed-off-by: Tim Chang <changti@amazon.com>

### Description

Adding support for FORMAT(). This is a slight rework of Samanwitha's original commits, linked below. I have squashed the commits, moved the format code into a separate file, restored the accidental reformatting of string.c, refactored parts of the code, and clarified comments and variable names.
 
### Task
BABEL-3286

### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.1.0--2.2.0.sql [hunk 1]
```diff
@@ -3584,6 +3584,75 @@ END;
 $$
 LANGUAGE 'pltsql';
 
+CREATE OR REPLACE FUNCTION sys.format_datetime(IN value anyelement, IN format_pattern NVARCHAR,IN culture VARCHAR,  IN data_type VARCHAR DEFAULT '') RETURNS sys.nvarchar
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.1.0--2.2.0.sql]: if this is targeted for Narwhal, change has to wait till we tag/branch manatee and create a new upgrade script

File: contrib/babelfishpg_tsql/sql/datatype_string_operators.sql [hunk 2]
```diff
@@ -56,3 +56,73 @@ GRANT EXECUTE ON FUNCTION sys.formatmessage(IN TEXT, VARIADIC "any") TO PUBLIC;
...
+    v_temp_integer INTEGER;
+BEGIN
+    arg_type := pg_typeof(arg);
+
+    CASE
+        WHEN arg_type IN ('time'::regtype ) THEN
+            RETURN sys.format_datetime(arg, p_format_pattern, p_culture, 'time');
+
+        WHEN arg_type IN ('date'::regtype, 'sys.datetime'::regtype, 'sys.smalldatetime'::regtype, 'sys.datetime2'::regtype ) THEN
+            RETURN sys.format_datetime(arg::timestamp, p_format_pattern, p_culture);
```
Comment (2jungkook) [inline contrib/babelfishpg_tsql/sql/datatype_string_operators.sql:83]: To minimize future confusion/mixed-up, let's use precise types, e.g., `sys.varchar`, `pg_catalog.timestamp`, etc.

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.1.0--2.2.0.sql [hunk 3]
```diff
@@ -3584,6 +3584,75 @@ END;
...
+EXCEPTION
+	WHEN datatype_mismatch THEN
+		RAISE USING MESSAGE := format('Argument data type % is invalid for argument 1 of format function.', pg_typeof(arg)),
+					DETAIL := 'Invalid datatype.',
+					HINT := 'Convert it to valid datatype and try again.';
+END;
+$BODY$
+LANGUAGE plpgsql IMMUTABLE PARALLEL SAFE;
+GRANT EXECUTE ON FUNCTION sys.FORMAT(IN anyelement, IN NVARCHAR, IN VARCHAR) TO PUBLIC;
+
```
Comment (2jungkook) [inline contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.1.0--2.2.0.sql]: You need to add other changes to the upgrade script as well, e.g., changes in `sys_function_helpers.sql`.

File: contrib/babelfishpg_tsql/src/format.c [hunk 4]
```diff
@@ -0,0 +1,2031 @@
...
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("The culture parameter \"%s\" provided in the function call is not supported.", culture),
+				 errhint("Invalid/Unsupported culture value.")));
+	}
+
+	arg_type_oid = get_fn_expr_argtype(fcinfo->flinfo, 0);
+
+	buf = palloc(4000 * sizeof(char));
```
Comment (2jungkook) [inline contrib/babelfishpg_tsql/src/format.c]: I saw many patterns using `palloc` and `strncpy` which has some dangerous quirks. Let's try to use `StringInfoData` unless such low-level functions are really necessary.

File: contrib/babelfishpg_tsql/src/format.c [hunk 5]
```diff
@@ -0,0 +1,2031 @@
...
+	VarChar *result;
+
+	if (PG_ARGISNULL(0))
+		PG_RETURN_NULL();
+
+	culture = text_to_cstring(PG_GETARG_TEXT_P(2));
+
+	valid_culture = format_validate_and_culture(culture, "LC_TIME");
+
+	if (strlen(valid_culture) <= 0)
```
Comment (2jungkook) [inline contrib/babelfishpg_tsql/src/format.c]: Can we move this invalid check into `format_validate_and_culture` function?

File: contrib/babelfishpg_tsql/src/format.c [hunk 6]
```diff
@@ -0,0 +1,2031 @@
...
+				 errhint("Invalid/Unsupported culture value.")));
+	}
+
+	arg_type_oid = get_fn_expr_argtype(fcinfo->flinfo, 0);
+
+	buf = palloc(4000 * sizeof(char));
+
+	if (PG_ARGISNULL(1))
+	{
+		PG_RETURN_NULL();
```
Comment (2jungkook) [inline contrib/babelfishpg_tsql/src/format.c:70]: Is this an expected result or an exceptional case?

File: contrib/babelfishpg_tsql/src/format.c [hunk 7]
```diff
@@ -0,0 +1,2031 @@
...
+		break;
+	}
+}
+
+
+/*
+ * Function for setting validated input locales for LC_TIME, LC_NUMERIC, LC_MONETARY
+ */
+int
+set_culture(char *valid_culture, const char *config_name, char *culture)
```
Comment (2jungkook) [inline contrib/babelfishpg_tsql/src/format.c:372]: `culture` argument is not used.

File: contrib/babelfishpg_tsql/src/format.c [hunk 8]
```diff
@@ -0,0 +1,2031 @@
...
+{
+	if (valid_culture != NULL && strlen(valid_culture) > 0)
+	{
+		if (pg_strcasecmp(config_name, "LC_MONETARY") != 0)
+			strncat(valid_culture, ".UTF-8", 7);
+
+		set_config_option(config_name, valid_culture,
```
Comment (2jungkook) [inline contrib/babelfishpg_tsql/src/format.c:381]: What is the life cycle of this config? Can it handle queries like
```
select format(..., 'en-US', format(..., 'fr-BE')
```

File: contrib/babelfishpg_tsql/src/format.c [hunk 9]
```diff
@@ -0,0 +1,2031 @@
...
+}
+
+/*
+ * Base function for getting the standard format mask for date and time data types
+ * Returns  0 if the format is invalid
+ * Returns -1 if the format is not supported
+ * Returns -2 if the culture couldn't be found to get the format
+ */
+int
+format_datetimeformats(char *buf, char *format_pattern, char *culture, char *data_type, char *data_val)
```
Comment (2jungkook) [inline contrib/babelfishpg_tsql/src/format.c]: Let's use `const char *` as much as possible if an argument is not changed inside a function.

File: contrib/babelfishpg_tsql/src/format.c [hunk 10]
```diff
@@ -0,0 +1,2031 @@
...
+get_currency_sign_format(char *culture, int positive)
+{
+	int flag = 0;
+	int j = 0;
+
+	for (j = 0; j < CULTURE_COUNT && flag == 0; j++)
+	{
+		if (pg_strcasecmp(currencyformats[j].sql_culture, culture) == 0)
+		{
+			flag = 1;
```
Comment (2jungkook) [inline contrib/babelfishpg_tsql/src/format.c]: This is unnecessary.

File: contrib/babelfishpg_tsql/src/format.c [hunk 11]
```diff
@@ -0,0 +1,2031 @@
...
+			{
+				return currencyformats[j].positive_pattern;
+			}
+			else
+			{
+				return currencyformats[j].negative_pattern;
+			}
+		}
+	}
+	return "";
```
Comment (2jungkook) [inline contrib/babelfishpg_tsql/src/format.c:1254]: What is the expected outcome for returning an empty string?

File: contrib/babelfishpg_tsql/src/format.c [hunk 12]
```diff
@@ -0,0 +1,2031 @@
...
+	int precision = get_precision(toupper(pattern), precision_string, "", 0, "");
+
+	format_exponential_helper(format_res, numeric_val, precision);
+
+	if (isupper(pattern))
+		for (char *c = format_res; *c; ++c) *c = toupper(*c);
+
+	len = strlen(format_res);
+
+	if (format_res[len - 5] != 'E' && format_res[len - 5] != 'e')
```
Comment (2jungkook) [inline contrib/babelfishpg_tsql/src/format.c]: Can we guarantee `len` is greater than or equal to 5?

File: contrib/babelfishpg_tsql/src/format.h [hunk 13]
```diff
@@ -0,0 +1,1775 @@
+#ifndef FORMAT_H
+#define FORMAT_H
+
+#include "postgres.h"
+#include "utils/numeric.h"
+
+#define CULTURE_COUNT (843)
```
Comment (2jungkook) [inline contrib/babelfishpg_tsql/src/format.h]: Can we avoid the hard-coded value by using the size of the struct?

File: test/JDBC/input/functions/format-vu-verify.sql [hunk 14]
```diff
@@ -0,0 +1,95 @@
+select FORMAT(d, 'd','en-us') from date_testing;
```
Comment (2jungkook) [inline test/JDBC/input/functions/format-vu-verify.sql:1]: One of the main purposes of MVU test is to verify dependent objects such as views functions and procedures. I would suggest to add at least the followings:
- `create view format_v1 as select FORMAT(...), ...`
- `create function format_f1 (... default FORMAT(...))...` (if possible)

**Comment by timchang514** [comment]
Currently blocked on adding upgrade script changes until the Narwhal branch is cut, so focusing on improving the C code.

File: contrib/babelfishpg_tsql/src/format.h [hunk 15]
```diff
@@ -0,0 +1,1779 @@
...
+	char *pg_shorttimepattern;
+	char *pg_longtimepattern;
+	char *pg_monthdaypattern;
+	char *pg_yearmonthpattern;
+} datetimeformat;
+
+/*
+ * Culture wise standard date/time format masks
+ */
+static const datetimeformat datetimeformats[] = {
```
Comment (forestkeeper) [inline contrib/babelfishpg_tsql/src/format.h:26]: Let's add a refer for this table

**Review by macwoody — DISMISSED** [review]

**Review by forestkeeper — DISMISSED** [review]

**Review by macwoody — APPROVED** [review]

**Pull request merged**